[CORE-12155] Introduce external timeout for cloud_storage client leases

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds external timeouts governing the various HTTP request originating from `cloud_io::remote`. The context here is that stream interfaces don't provide any timeout facility, so a stream wrapping a connected `http::client` can hang indefinitely if the underlying connection gets stuck.

General idea:
1. Embed a watchdog timer in each `client_lease` acquired from `client_pool`
2. If the watchdog fires before the lease is returned, call `client::shutdown` immediately to cleanly terminate any open connections and log a message.

TODO:
- [x] Add some more background to this comment
- [x] ducktape test is hardcoded to to the IP address of my local minio container. replace with something generic that will work across docker & CDT
- [ ] is this sufficiently well tested?
   - [x] are we confident it doesn't break things when timeouts occur? (I think so)
   - [x] are we confident that it gets us out of jail during a stuck connection? (??)

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
cio/remote: Use acquire_with_timeout for client leases

Pass the static logger and rtc context for output.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dt/e2e_si: Traffic shaping test

General idea here is to trigger timeouts in 'cloud_io::remote' by adding
delays to traffic from the redpanda node -> s3. To that end, add a stub
node interface that wraps the IP address of the s3 service and use
ClusterTopology as usual.

Note that we have no way to shape the other direction of the link (s3 -> RP)

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
csc: client_pool::acquire_with_timeout

Takes a 'duration' and a 'logger' and produces a 'client_lease'. This client
lease embeds a watchdog whose timeout handler logs a message and shuts down
its client.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1

oleiman commented:
/ci-repeat 1

oleiman commented:
CI Failures: watchdog handler should not hold a reference to local `retry_chain_logger`

oleiman commented:
/ci-repeat 1

Lazin commented:
this looks much better, a localised change

oleiman commented:
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingManyPartitionsTest
dt-repeat=25

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67350-->
<!--94eac3ec-8646-4dc0-8163-8f24937bc9ef-->
<!--title-->
## Retry command for [Build#67350](https://buildkite.com/redpanda/redpanda/builds/67350)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--536027f6-3bdc-432f-bdab-c542072eabcc-->
<!--TEST RESULTS-->
<!--f3ac2e8a-65ff-4da8-8175-d8208ecee916-->
<!--TEST RESULTS-->
<!--9d3bf939-7af3-42a7-b974-2d7d44ccfee8-->
<!--TEST RESULTS-->
<!--2af7c21c-18d1-44af-8c33-362a737e9a6a-->
<!--TEST RESULTS-->
<!--6dd26df5-fd51-49b6-92e4-8fcff47210da-->
<!--TEST RESULTS-->
<!--0d9f46f0-e566-4129-ba3f-31ec6db9b052-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67350">build#67350</a></summary>


| test_class                                   | test_method                                        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                    |
|:---------------------------------------------|:---------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------|
| ShadowIndexingTrafficShapingTest             | test_many_partitions_recovery_with_traffic_shaping |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67350#01976b7b-f5fe-421e-a4fd-e31c7ef65ada | FLAKY         | 13/45    | The test was found to be new, and no failures are allowed |
| src/v/crypto/tests/crypto_bench_rpbench_test | src/v/crypto/tests/crypto_bench_rpbench_test       |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/67350#01976b4b-0529-4c1a-a34a-263cba658648 | FAIL          | 0/1      |                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67452">build#67452</a></summary>


| test_class                    | test_method                               | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                               |
|:------------------------------|:------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67452#01977cec-78df-461c-aa4c-d09b56ba7439 | FLAKY         | 19/21    | upstream reliability is '93.4065934065934'. current run reliability is '90.47619047619048'. drift is 2.9304 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67513">build#67513</a></summary>


| test_class                        | test_method                               | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest     | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978005-1b49-4a4c-b5c5-2fa8ae0a516b | FLAKY         | 15/21    | upstream reliability is '93.43629343629344'. current run reliability is '71.42857142857143'. drift is 22.00772 and the allowed drift is set to 50. The test should PASS |
| PartitionForceReconfigurationTest | test_basic_reconfiguration                | {"acks": -1, "controller_snapshots": false, "restart": false}                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978028-dd0b-452d-bfb4-32bb744581a7 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RackAwarePlacementTest            | test_replica_placement                    | {"num_partitions": 50, "num_topics": 2, "rack_layout_str": "xxYYzz", "replication_factor": 3}                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978028-dd0b-4bea-a25c-1ab3b1c0c3bf | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest          | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978005-1b4b-4415-9f5b-75cf8821de70 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             |
| RandomNodeOperationsTest          | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978005-1b47-4448-94d0-79b44d4bec42 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| TopicDeleteCloudStorageTest       | drop_lifecycle_marker_test                | {"cloud_storage_type": 2}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67513#01978028-dd0d-4fbf-82c8-2bb1d7b87156 | FLAKY         | 19/21    | upstream reliability is '97.7124183006536'. current run reliability is '90.47619047619048'. drift is 7.23623 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67583">build#67583</a></summary>


| test_class             | test_method            | test_arguments                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------|:-----------------------|:-------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| TestTieredStoragePause | test_safe_pause_resume | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67583#01978501-816d-4b59-8b99-f8b136948ae0 | FLAKY         | 20/21    | upstream reliability is '90.93484419263456'. current run reliability is '95.23809523809523'. drift is -4.30325 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67630">build#67630</a></summary>


| test_class                   | test_method                                       | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:--------------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves           | {"cleanup_policy": "compact,delete"}                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67630#0197899f-9df5-47ab-9203-c9c9f27a6fe8 | FLAKY         | 19/21    | upstream reliability is '97.9933110367893'. current run reliability is '90.47619047619048'. drift is 7.51712 and the allowed drift is set to 50. The test should PASS   |
| DatalakeE2ETests             | test_latest_protobuf_schema                       | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67630#01978a66-ef88-4e41-9903-605bc42e1083 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionBalancerTest        | test_fuzz_admin_ops                               |                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67630#0197899f-9df3-4df6-92e1-84c93c3c3bf0 | FLAKY         | 20/21    | upstream reliability is '96.34703196347031'. current run reliability is '95.23809523809523'. drift is 1.10894 and the allowed drift is set to 50. The test should PASS  |
| PartitionReassignmentsTest   | test_add_partitions_with_inprogress_reassignments |                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67630#0197899f-9df4-4290-a758-92aae9c9655c | FLAKY         | 16/21    | upstream reliability is '93.3641975308642'. current run reliability is '76.19047619047619'. drift is 17.17372 and the allowed drift is set to 50. The test should PASS  |
| RaftAvailabilityTest         | test_controller_node_isolation                    |                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67630#01978a66-ef8a-4ee7-9173-57ec683f5edd | FLAKY         | 20/21    | upstream reliability is '94.52380952380952'. current run reliability is '95.23809523809523'. drift is -0.71429 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67697">build#67697</a></summary>


| test_class                          | test_method                      | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:------------------------------------|:---------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| archival_metadata_stm_gtest_fixture | test_same_term_sync_abort_source |                                                                                                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/67697#019799f6-35d5-4a71-8dfb-29d2c3ec3c1c | FAIL          | 0/1      |                                                                                                                                                           |
| RandomNodeOperationsTest            | test_node_operations             | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67697#01979a25-ad79-4591-9e8e-e24bf9547dad | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/e143e43cf75b45602d3ad718e08eb996e7508f79..e6fe9cb5bd57d703836de8c2186953b9914c55b0) to drop expected number of segment uploads expected in the ducktape test by half, since we're degrading network perf pretty significantly.

oleiman commented:
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

dotnwat commented on src/v/ssx/watchdog.h:
```diff
@@ -90,8 +99,13 @@ class watchdog {
     }
-    void cancel() { _as.request_abort(); }
+    void cancel() {
+        if (!moved_from) {
```
it kinda looks like it should be safe to call `request_abort` on a default-constructed abort source? or, did you add the `moved_from` because `std::exception_ptr` isn't cleared by std::move (std::exchange should fix that)?

oleiman replied:
the latter. 

> std::exchange should fix that

yeah, it's a good point.

oleiman replied:
updated and subsequently dropped this commit.

dotnwat commented on src/v/ssx/watchdog.h:
```diff
@@ -77,9 +77,18 @@ class watchdog {
     watchdog(const watchdog&) = delete;
-    watchdog(watchdog&&) = delete;
+    watchdog(watchdog&& other) noexcept
+      : _as(std::move(other._as))
```
prefer something like `_as(std::exchange(other._as, {}))` to make it clear that the other abort source should no longer be active.

oleiman replied:
done

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/05e89519fb73fd63dae1de5323c2e08d29528d84..5018870d8e6e3b812c1c03166656e6048be50b04) use `std::exchange` in watchdog move ops.

Lazin commented on src/v/ssx/watchdog.h:
```diff
@@ -77,9 +77,13 @@ class watchdog {
     watchdog& operator=(const watchdog&) = delete;
-    watchdog& operator=(watchdog&&) = delete;
+    watchdog& operator=(watchdog&& other) noexcept {
+        _as = std::exchange(other._as, {});
```
I think that there is a problem with this. The `_as` is moved and then in d-tor the `_as.request_abort()` is invoked. So basically the method of the moved object is invoked which is an UB.

oleiman replied:
In this arrangement, `other._as` should be left in a default-initialized state, or am I missing something obvious?

oleiman replied:
anyway i dropped this commit per reminder that pointers exist 🙃

Lazin commented on src/v/cloud_storage_clients/client_pool.h:
```diff
@@ -128,6 +132,47 @@ class client_pool
+      ss::lowres_clock::duration to,
+      Logger& log,
+      std::optional<ss::sstring> ctx = std::nullopt) {
+        auto lease = co_await acquire(as);
```
This doesn't have a timeout. If for some reason all clients are acquired and never returned this call could get stuck here. So maybe it makes sense to add an optional 'timeout' parameter to the 'acquire' method?

Lazin replied:
The problem is very unlikely but still. If the method gets timeout as an input it'd be easier to reason about it if it's guaranteed that the timeout is respected.

oleiman replied:
good idea yeah

oleiman replied:
some definition of done. set it up so the acquire calls in `cloud_io::remote` just cause timeout/retry error if failed.

Lazin commented on src/v/cloud_storage_clients/client_pool.h:
```diff
@@ -128,6 +132,47 @@ class client_pool
+    ss::future<client_lease> acquire_with_timeout(
+      ss::abort_source& as,
+      ss::lowres_clock::duration to,
+      Logger& log,
```
I think it's OK to use the logger of the pool for this. Without the parameter you can move this method to the .cc file.

oleiman replied:
yeah. originally had wanted to push the rtc logger through, but that complicates lifetime management.

oleiman replied:
done

Lazin commented on src/v/cloud_storage_clients/client_pool.h:
```diff
@@ -128,6 +132,47 @@ class client_pool
+        auto lease = co_await acquire(as);
+        // take a copy of the shared_ptr held by the lease to avoid racing with
+        // client_pool teardown
+        lease._wd.emplace(
```
Maybe the deleter of the lease is a better place to allocate the watchdog? It could be one of the captured parameters of the lambda. This way it doesn't have to be moveable. Or maybe this could be a unique_ptr.

oleiman replied:
> `unique_ptr`

yeah fair. deleter feels a bit fussy tbh but I'll play with it.

oleiman replied:
done w/ `unique_ptr`, which seems much simpler

oleiman committed:
csc: Add an optional timeout to client_pool::acquire

So we don't wait forever when/if the pool is fully subscribed for a long
period of time.

oleiman commented:
/ci-repeat 1

mmaslankaprv commented on src/v/cluster/cloud_metadata/tests/manifest_downloads_test.cc:
```diff
@@ -59,7 +59,7 @@ TEST_F(cluster_metadata_fixture, test_download_manifest) {
     // the first time we're using the bucket for this cluster.
     retry_chain_node retry_node(
-      never_abort, ss::lowres_clock::time_point::max(), 10ms);
+      never_abort, ss::lowres_clock::now() + 1h, 10ms);
```
nit: maybe define this as a constant ?

oleiman replied:
dropped. passing max is once again safe to do.

mmaslankaprv commented on src/v/utils/retry_chain_node.cc:
```diff
@@ -300,6 +300,10 @@ basic_retry_chain_node<Clock>::get_poll_interval() const {
 
 template<class Clock>
 typename Clock::duration basic_retry_chain_node<Clock>::get_timeout() const {
+    vassert(
```
i am wondering if this can be a `static_assert`, this way we would be warned at compile time

oleiman replied:
i don't think the deadline is generally known at compile time. am i missing something?

oleiman replied:
i don't really think this is a good place for the assertion anyway. may just move it to `client_pool`

oleiman replied:
ditched all of this and put a conditional guard in client pool to avoid hitting overflows in watchdog.

mmaslankaprv commented on src/v/cloud_storage_clients/client_pool.cc:
```diff
@@ -426,6 +426,33 @@ client_pool::acquire(ss::abort_source& as) {
+    auto lease = co_await acquire(as);
+    // take a copy of the shared_ptr held by the lease to avoid racing with
+    // client_pool teardown
+    auto to = deadline - ss::lowres_clock::now();
```
nit: would it be helpful to expose `watchdog` constructor accepting `time_point` ?

oleiman replied:
maybe so. for the logging, i think the `duration` is more helpful than the `time_point`, so it feels like the same same.

mmaslankaprv commented on tests/rptest/tests/e2e_shadow_indexing_test.py:
```diff
@@ -1552,3 +1563,155 @@ def test_hydration_completes_when_consumer_killed(self):
+            return len(
+                [s for s in self.log_searcher._capture_log(node, f"'{msg}'")])
+
+        with ClusterTopology() as topology:
```
i think we may not need the whole topology setup here, it should be enough to add timeouts using a raw `NodeQdisc` ? Also to match incoming trafic you may consider adding this to the `NodeQdisc`:
```python
    def drop_incoming(self, src_address):
        self._tc_execute(
            ['qdisc', 'add', 'dev', self.dev, 'handle', 'ffff:', 'ingress'])
        self._tc_execute([
            'filter', 'add', 'dev', self.dev, 'parent', 'ffff:', 'protocol',
            'ip', 'u32', 'match', 'ip', 'src', src_address, "action", "drop"
        ])
```

oleiman replied:
> may not need the whole topology setup here

yeah that's a fair point

> match incoming traffic

nice one. thanks!

oleiman replied:
done. much more clear IMO for a limited usage case.

Lazin reviewed (dismissed):
LGTM

oleiman commented:
Note to reviewers: Should have another CDT run w/ the additional shaping rules

oleiman commented:
CI Failure:
- [build-debug](https://buildkite.com/redpanda/redpanda/builds/67583#019784b0-3911-43ab-af18-e8dc3ec2f44d) looks spurious. don't see an actual test failure there.

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=50

dotnwat commented on src/v/cloud_storage_clients/client_pool.cc:
```diff
@@ -426,6 +426,35 @@ client_pool::acquire(ss::abort_source& as) {
+        // take a copy of the shared_ptr held by the lease to avoid racing with
+        // client_pool teardown
+        auto to = deadline - ss::lowres_clock::now();
+        lease._wd = std::make_unique<ssx::watchdog>(
```
i don't know much about the client pool. is a lease 1-1 / exclusive with a client instance?

oleiman replied:
yea that's how i understood it. there's a concept of "borrowing" but i believe that's grabbing a free client from another shard?

dotnwat commented on src/v/cloud_storage_clients/client_pool.cc:
```diff
@@ -426,6 +426,35 @@ client_pool::acquire(ss::abort_source& as) {
+    if (deadline < ss::lowres_clock::time_point::max()) {
+        // take a copy of the shared_ptr held by the lease to avoid racing with
+        // client_pool teardown
+        auto to = deadline - ss::lowres_clock::now();
```
what happens if `to` is negative? seems like with fib.get_deadline and all the co_awaits that it's at least possible that some caller might pass in a deadline that has already been reached?

oleiman replied:
ugh, good catch

oleiman replied:
looks safe to me. added a couple unit tests

dotnwat replied:
this looks the same, right? did you mean that it looks safe in its unmodified form as in it can't be negative or if its negative it still does the right thing?

WillemKauf replied:
I believe his unit test shows its negative, it still does the right thing: https://github.com/redpanda-data/redpanda/pull/26456/commits/0d83d6158eb612f6b92b7cdc6ec40ad2f95e2c70

oleiman replied:
yea, if it's negative it does the right thing, which is to fire async immediately. no special handling because we're just bonking the client in the handler.

dotnwat commented on src/v/cloud_io/remote.cc:
```diff
@@ -85,6 +85,17 @@ cloud_io::provider infer_provider(
+    } catch (const ss::timed_out_error&) {
+        return on_timeout;
+    } catch (...) {
+        std::rethrow_exception(std::current_exception());
```
nit: i think just `throw;` is the short hand for this

oleiman replied:
fixed

dotnwat commented on src/v/cloud_io/remote.cc:
```diff
@@ -243,8 +254,14 @@ ss::future<upload_result> remote::upload_stream(
         }
-        auto lease = co_await _pool.local().acquire_with_timeout(
-          fib.root_abort_source(), fib.get_deadline(), fib());
+        auto fut = co_await ss::coroutine::as_future(
+          _pool.local().acquire_with_timeout(
+            fib.root_abort_source(), fib.get_deadline(), fib()));
+        if (fut.failed()) {
```
is the expectation that acquire_with_timeout will cause `fut` to be failed if the watchdog fails? if so, i'm not quite seeing how that works. it looks like acquire_with_timeout returns the lease as soon as it's acquired and that happens independent of the outcome of the watchdog that is setup?

oleiman replied:
no, the acquisition of the lease is basically a synchronous loop that will throw if the deadline is reached before a client becomes available.

the watchdog is set after the client pool hands out a lease, and the intent is that the deadline will trigger a forced shutdown on the client, killing any connections and winding back out to the request site w/ a non -exceptional failure.

sounds like this isn't very clear from context. i'll update the comments.

oleiman replied:
done

oleiman committed:
ssx/tests: Test that ssx::watchdog accepts a negative timeout

Watchdog fires, passes sanitizers

WillemKauf reviewed (dismissed):
Really cool stuff, great changes

WillemKauf commented on src/v/cloud_storage_clients/client_pool.cc:
```diff
@@ -426,6 +426,35 @@ client_pool::acquire(ss::abort_source& as) {
+                    "Lease expired after {}. Shutting down client...",
+                    to);
+              }
+              client->shutdown();
```
Maybe we want a probe here? some sort of metric that tells us how many clients are timing out to help detect instability in the cloud storage layer?

(this is probably fine as a follow up if we decide its beneficial)

oleiman replied:
good idea. done

oleiman committed:
csc: Add timeout counter to client_probe

Called when the lease watchdog expires.

'cloud_client_lease_timeouts_total'

oleiman commented:
CI Failure:
- [build-debug](https://buildkite.com/redpanda/redpanda/builds/67697#019799f6-35d5-4a71-8dfb-29d2c3ec3c1c) `archival_metadata_stm_gtest` flake

oleiman commented:
CI Failure:
- [release-duck](https://buildkite.com/redpanda/redpanda/builds/67697#01979a25-ad7b-4903-b219-ba9915ff02a8) all passed, looks like infra

oleiman commented:
/cdt
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingTrafficShapingTest.test_many_partitions_recovery_with_traffic_shaping
dt-repeat=25

mmaslankaprv commented:
looks great, thanks for adding that `NodeQdisc` method 👍🏼

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26456-v24.2.x-410 remotes/upstream/v24.2.x
git cherry-pick -x 8467da807d 0d83d6158e b29112efa1 f67718950d 18a5decb7e a7bcb801df
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/15826303755).

WillemKauf commented on src/v/cloud_storage_clients/client_pool.cc:
```diff
@@ -456,6 +460,7 @@ auto client_pool::acquire_with_timeout(
                     "Lease expired after {}. Shutting down client...",
                     to);
               }
+              probe->register_timeout();
```
Didn't think to call this out, but might this need a check for `if(probe)`? 
Might race with `client_pool::stop()` which sets it to `nullptr`.

oleiman replied:
good shout. we capture a copy of the shared pointer, so it looks safe to me, lifetime-wise. looking at `cloud_storage_clients/configuration`, it appears that we `make_shared<client_probe>` unconditionally, so no worry there I think.

Can still put up a diff to check, but I don't think it's pants-on-fire urgent.

oleiman replied:
https://github.com/redpanda-data/redpanda/pull/26566
